scripts/templates/params.hpp.mako

```diff
@@ -0,0 +1,233 @@
+            %endfor
+                default:
+                    os << (value == 0 ? "0" : "invalid enumerator");
+                    break;
+        }
+        return os;
+    }
+    %endif
+    %if "structure_type" in obj['name']:
+    inline void serializeStruct(std::ostream &os, const void *ptr) {
+        if (ptr == NULL) {
+            serializePtr(os, ptr);
+            return;
+        }
+
+        ## structure type enum value must be first
+        enum ${th.make_enum_name(n, tags, obj)} *value = (enum ${th.make_enum_name(n, tags, obj)} *)ptr;
+        switch (*value) {
+            %for n, item in enumerate(obj['etors']):
+                <%
+                ename = th.make_etor_name(n, tags, obj['name'], item['name'])
+                %>
+                case ${ename}: {
+                    const ${th.subt(n, tags, item['desc'])} *pstruct = (const ${th.subt(n, tags, item['desc'])} *)ptr;
+                    serializePtr(os, pstruct);
+                } break;
+            %endfor
+                default:
+                    os << "invalid enumerator";
+                    break;
+        }
+    }
+    %endif
+%if th.type_traits.is_flags(obj['name']):
+inline void serializeFlag_${th.make_type_name(n, tags, obj)}(std::ostream &os, ${th.make_type_name(n, tags, obj)} flag) {
+    uint32_t val = flag;
+    bool first = true;
+    %for n, item in enumerate(obj['etors']):
+        <%
+        ename = th.make_etor_name(n, tags, obj['name'], item['name'])
+        %>
+        if ((val & ${ename}) == (uint32_t)${ename}) {
+            ## toggle the bits to avoid printing overlapping values
+            ## instead of e.g., FLAG_FOO | FLAG_BAR | FLAG_ALL, this will just
+            ## print FLAG_FOO | FLAG_BAR (or just FLAG_ALL, depending on order).
+            val ^= (uint32_t)${ename};
+            if (!first) {
+                os << " | ";
+            } else {
+                first = false;
+            }
+            os << ${ename};
+        }
+    %endfor
+    if (first) {
+        os << "0";
+    }
+}
+%endif
+## STRUCT/UNION ###############################################################
+%elif re.match(r"struct|union", obj['type']):
+inline std::ostream &operator<<(std::ostream &os, const ${obj['type']} ${th.make_type_name(n, tags, obj)} params) {
+    os << "(${obj['type']} ${th.make_type_name(n, tags, obj)}){";
+    <% mlist = obj['members'] %>
+    %for n, item in enumerate(mlist):
+        ${line(item, n, False)}
+    %endfor
+    os << "}";
+    return os;
+}
+%endif
+%endfor # obj in spec['objects']
+%endfor
+
+%for tbl in th.get_pfncbtables(specs, meta, n, tags):
+%for obj in tbl['functions']:
+
+inline std::ostream &operator<<(std::ostream &os, const struct ${th.make_pfncb_param_type(n, tags, obj)} *params) {
+    %for n, item in enumerate(obj['params']):
+        ${line(item, n, True)}
+    %endfor
+    return os;
+}
+
+%endfor
+%endfor
+
+template <typename T> inline void serializePtr(std::ostream &os, T *ptr) {
```

Comment:
Do we need to have those helper functions in `mako`? I think it would be better to put generic functions directly to some .hpp file